examples/apps/ai-collab/src/components/TaskGroup.tsx

```diff
@@ -59,8 +71,94 @@ export function TaskGroup(props: {
 		newBranchTargetNode: SharedTreeTaskGroup;
 	}>();
 
+	const [undoStack, setUndoStack] = useState<Revertible[]>([]);
+	const [redoStack, setRedoStack] = useState<Revertible[]>([]);
+
 	useSharedTreeRerender({ sharedTreeNode: props.sharedTreeTaskGroup, logId: "TaskGroup" });
 
+	/**
+	 * Create undo and redo stacks of {@link Revertible}.
+	 */
+	useEffect(() => {
+		function onRevertibleDisposed(disposed: Revertible): void {
+			const redoIndex = redoStack.indexOf(disposed);
+			if (redoIndex === -1) {
+				const undoIndex = undoStack.indexOf(disposed);
+				if (undoIndex !== -1) {
+					setUndoStack((currUndoStack) => {
+						const newUndoStack = [...currUndoStack];
+						newUndoStack.splice(undoIndex, 1);
+						return newUndoStack;
```

Comment:
Nit, reading the documentation to refresh myself on `splice` I saw that we can do this in a single step. It returns a new array, so it'll still play nice with React.
```suggestion
						const newUndoStack = currUndoStack.toSpliced(undoIndex, 1);
						return newUndoStack;
```